common.mk

```diff
@@ -63,37 +63,23 @@ SCALA_SOURCES = $(call lookup_srcs,$(SOURCE_DIRS),scala)
 VLOG_SOURCES = $(call lookup_srcs,$(SOURCE_DIRS),sv) $(call lookup_srcs,$(SOURCE_DIRS),v)
 # This assumes no SBT meta-build sources
 SBT_SOURCE_DIRS = $(addprefix $(base_dir)/,generators sims/firesim/sim tools)
-SBT_SOURCES = $(call lookup_srcs,$(SBT_SOURCE_DIRS),sbt) $(base_dir)/build.sbt $(base_dir)/project/plugins.sbt
+SBT_SOURCES = $(call lookup_srcs,$(SBT_SOURCE_DIRS),sbt) $(base_dir)/build.sbt $(base_dir)/project/plugins.sbt $(base_dir)/project/build.properties
 
 #########################################################################################
-# jar creation variables and rules
+# SBT Server Setup (needed to rebuild project correctly)
 #########################################################################################
-FIRRTL_JAR := $(base_dir)/lib/firrtl.jar
-FIRRTL_TEST_JAR := $(base_dir)/test_lib/firrtl-test.jar
-
-$(FIRRTL_JAR): $(call lookup_srcs,$(CHIPYARD_FIRRTL_DIR),scala)
-	$(MAKE) -C $(CHIPYARD_FIRRTL_DIR) SBT="$(SBT)" root_dir=$(CHIPYARD_FIRRTL_DIR) build-scala
-	mkdir -p $(@D)
-	cp -p $(CHIPYARD_FIRRTL_DIR)/utils/bin/firrtl.jar $@
-	touch $@
-
-$(FIRRTL_TEST_JAR): $(call lookup_srcs,$(CHIPYARD_FIRRTL_DIR),scala)
-	cd $(CHIPYARD_FIRRTL_DIR) && $(SBT) "test:assembly"
-	mkdir -p $(@D)
-	cp -p $(CHIPYARD_FIRRTL_DIR)/utils/bin/firrtl-test.jar $@
+$(SBT_THIN_CLIENT_TIMESTAMP): $(SBT_SOURCES)
```

Comment:
This is a bit hacky.

If you don't do this, changing the `build.sbt` would require you to `$(SBT) shutdown` the server then continue with the normal workflow (i.e. run `make verilog` or whatever afterwards). Doing this would re-spin up the thin server and reload all the project definitions. 

Once the https://github.com/sbt/sbt/issues/4638 issue in SBT is resolved, we can avoid this `ifneq` logic by just doing:

```
cd $(base_dir) && $(SBT) "shutdown"
touch $@
```

The reason I don't do this now is that doing `$(SBT) shutdown` the 1st time using the thin client will spin-up the server, then kill it resulting in the next `$(SBT)` call spinning-up the server again (unnecessary loading project def.s twice on both spinups).

Reply:
Is there no means to send a `reload` command to the server?

Reply:
Actually, you don't want to send a `reload` in every case that the `build.sbt` is changed. For the same reason mentioned in the last paragraph. Sending a `reload` before the server is started will read the proj. def.s twice. Hence I "hack" around this by checking if the `SBT_TIMESTAMP` file exists to avoid sending an extra `reload` when the server is down (which is most likely when the `SBT_TIMESTAMP` file doesn't exist... though not always the case).

Here is timeline of events in the **good** case (what I do):
1. TIMESTAMP file doesn't exist
2. `make ...`
3. Create TIMESTAMP file WITHOUT running `reload`
4. First SBT call is done (to `GenSimFiles`)... which reads the proj. def. and starts the server
5. Finished Verilog
6. Touch `build.sbt` and re-run `make ...`
7.  Update TIMESTAMP file and run `reload` to update the server

Here is timeline of events in the **bad** case (what I don't do):
1. TIMESTAMP file doesn't exist
2. `make ...`
3. Create TIMESTAMP file and run `reload` (This reads the proj. def.s twice, once for starting the server, once for `reload`). This would be the 1st SBT call.
4. 2nd SBT call is done (to `GenSimFiles`)... with server already started
5. Finished Verilog
6. Touch `build.sbt` and re-run `make ...`
7.  Update TIMESTAMP file and run `reload` to update the server

Reply:
I don't think this exactly what it was intended for but you could use the presence of the named-pipe
for the sbt server (https://www.scala-sbt.org/1.x/docs/sbt-server.html) to detect if it's running, You could instead touch `project/target/active.json` instead of creating and managing your own timestamp.

Reply:
That looks good to me and seems to pass a couple of quick scenarios. Good catch.

Reply:
Ok for future reference this is the event sequence when the server isn't started:

1. `active.json` doesn't exist
2. `make ...`
3. Run `sbt --client exit` to start server and quickly exit. (1st call)
4. Normal `make ...` sequence (`GenSimFiles`..., etc)
5. Touch `build.sbt`
6. Touch `active.json` and `sbt reload` to update the server.